### PR TITLE
Change serialized representation to unroll `And`.

### LINKED ISSUE #7: Maxing-out protobuf recursion depth
It appears that creating very large trees is causing transactions to fail. Originally I assumed this was due to GRPC message size limits, but even after @brian-weir increased this, we're still getting crashes.

It's been suggested that this is due to recursion limits in protobuf. See below forum post:

https://discuss.daml.com/t/sandbox-grpc-exception-failed-to-parse-submission/2595/7

Several suggestions have been discussed.

1. The least effort approach would be to 'flatten' `And` nodes, by redefining them as `| And [Claim]`. This doesn't fully solve the problem, but it does mitigate the depth, based on empirical cases. The other downside of this is that we're moving away from the original representation given in the papers, and while it's isomorphic to that representation, there may be unknowns we're currently not conscious of. 
2. Redefine `Serialized.Claim` to be a template; this would mean that instead of having a single contract holding a tree of claims, you would have one contract per node; the fixed-point would then be via a `ContractId _`, and this would break the recursion limit. The downside of this is that we lose generality of the asset parameter, `a`, until we get truly generic templates in the language. It also makes it difficult to work with `[de]serialize` as this will no longer be a pure function, but an `Update` [co]Kleisli.
A further disadvantage is that passing claims across the API would be very complicated: a client program would have to fetch each node of the tree individually and then reconstruct it off-ledger. Because nodes are recursively defined, in the worst-case, fetching an instrument would take O(n) whit n being the RTT.

### PR DESCRIPTION
**Pull request by lucianojoublanc-da**
That is, a right-recursive `And` structure will become a list of `Claim` in a
single `Serializable.And [Claim]`. This makes trees much more 'shallow' and resolves #7.

### CODE REVIEW

File: daml/ContingentClaims/Claim.daml [hunk 1]
```diff
@@ -114,29 +114,30 @@ isBearerExecutor = cata \case
 
 -- | Converts a `Serialized.Claim` read from ledger, to a more general `Claim`, for use with `Daml.Control.Recursion`.
 deserialize : Serialized.Claim t a -> Claim Observation t a
-deserialize = ana \case
+deserialize = futu \case
```
Comment (brian-weir) [inline daml/ContingentClaims/Claim.daml:117]: What does the change from `ana` to `futu` bring ?

File: daml/ContingentClaims/Claim.daml [hunk 2]
```diff
@@ -114,29 +114,30 @@ isBearerExecutor = cata \case
...
-serialize = ana \case
-  Zero -> Serialized.ZeroF
-  One a -> Serialized.OneF a
-  Give c -> Serialized.GiveF c
-  And c c' -> Serialized.AndF c c'
-  Or c c' -> Serialized.OrF c c'
-  Cond k c c' -> Serialized.CondF k c c'
-  Scale k c -> Serialized.ScaleF k c
-  When p c -> Serialized.WhenF p c
+serialize = cata \case
```
Comment (brian-weir) [inline daml/ContingentClaims/Claim.daml:132]: As above, why the change from `ana` to `cata` ?

**Review by brian-weir — APPROVED** [review]

File: daml/ContingentClaims/Claim.daml (hunk 1 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:117]: Hey, I added some doc to daml-ctl, so hopefully you can see some comments for these functions in your IDE now.

a `futumorphism` allows you to unfold more than a single level at a time. The first parameter is normally a `a -> f a` for `cata`, so there's no way to unfold more than a single `AndF` in one go (e.g. two levels would be `a -> f (f a)`); the type system prevents you from doing this. `futu` by contrast allows you to unfold multiple levels of the tree, safely.

File: daml/ContingentClaims/Claim.daml (hunk 2 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:132]: Both work, but it's more than anything for symmetry. If `deserialize` is an unfold (`ana`morphism), then `serialize` should be the dual, a fold (`cata`morphism). If you do this, you can fuse the two operations using a `hylo` (which combines `ana` followed by a `cata` - see type signature). You can still fuse passes if they're both folds, but it becomes trickier if you want to do other things in between, which is a typical pattern i.e. deserialize -> transform -> serialize.

File: daml/ContingentClaims/Claim.daml [hunk 3]
```diff
@@ -114,29 +114,30 @@ isBearerExecutor = cata \case
...
+  ZeroF -> Serialized.Zero
+  OneF a -> Serialized.One a
+  GiveF c -> Serialized.Give c
+  AndF c c' -> Serialized.And [c, c']
```
Comment (leonidr-da) [inline daml/ContingentClaims/Claim.daml:136]: Does this give us the "compression" that we're looking for? Should this be a `gcata` to have the current state of the translation?

**Review by leonidr-da — COMMENTED** [review]
Could we please add a test to demonstrate the conversion?

File: daml/ContingentClaims/Claim.daml (hunk 2 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml]: And funnily enough, I've now had to change this to `histo` (which is the dual of `futu` - not `cata`) for it to work.

File: daml/ContingentClaims/Claim.daml (hunk 3 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml]: Yes, but in fact this only covered the base (leaf) case. I've now fixed it (4ed4e27) to check cases where `c'` is also an `AndF`, and merge that into the current node.

Note that we've only changed `Serializable.Claim`, and not the (unserialized) `Claim`, so the DSL hasn't changed at all. But this does have an effect on the JSON API; the claims there will show the new representation.

File: daml/ContingentClaims/Claim.daml [hunk 4]
```diff
@@ -114,29 +114,31 @@ isBearerExecutor = cata \case
...
-  And c c' -> Serialized.AndF c c'
-  Or c c' -> Serialized.OrF c c'
-  Cond k c c' -> Serialized.CondF k c c'
-  Scale k c -> Serialized.ScaleF k c
-  When p c -> Serialized.WhenF p c
+serialize = histo \case
+  ZeroF -> Serialized.Zero
+  OneF a -> Serialized.One a
+  GiveF c -> Serialized.Give c.attribute
+  AndF c (Cofree (Serialized.And cs) _) -> Serialized.And $ c.attribute :: cs
```
Comment (lucianojoublanc-da) [inline daml/ContingentClaims/Claim.daml:136]: Thinking about this some more, this only covers the right-associative case. I'm hesitant to also merge the left subtree, as then `serialize . deserialize /= identity`. It would be weird for users if they serialize a tree, and then upon de-serialization it has been 're-balanced'. Let me know your thoughts.

Comment (leonidr-da) [reply daml/ContingentClaims/Claim.daml:136]: Can we define right associative ? `And (And 1 2) 3` won't become `And [1,2,3]`  but `And 1 (And 2 3)` will?

Regardless, I think this is the right (no pun intended) tradeoff of maintaining identity by providing compression on only one type of associativity.

Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:136]: > Can we define right associative ? And (And 1 2) 3 won't become And [1,2,3] but And 1 (And 2 3) will?

Exactly.

Mulling over it more, I do think that the breaking the identity `serialize . deserialize == identity` might be really counter-intuitive, so let's keep it like this.

I'm currently writing the tests for this and should be done shortly.

Comment (leonidr-da) [reply daml/ContingentClaims/Claim.daml:136]: Great

**Comment by lucianojoublanc-da** [comment]
> Could we please add a test to demonstrate the conversion?

Done. It highlighted some errors I had. I should probably writing more of these ...

File: daml/ContingentClaims/Claim.daml [hunk 5]
```diff
@@ -114,33 +114,41 @@ isBearerExecutor = cata \case
...
+deserialize = ana deserialize' 
+
+-- | F-algebra for `deserialize`; it can be composed in unfolds.
+deserialize' : Serialized.Claim t a -> ClaimF Observation t a (Serialized.Claim t a)
+deserialize' Serialized.Zero = ZeroF
+deserialize' (Serialized.One a) = OneF a
+deserialize' (Serialized.Give c) = GiveF c
+deserialize' (Serialized.And [c]) = deserialize'  c
+deserialize' (Serialized.And (c :: cs)) = AndF c (Serialized.And cs)
+deserialize' (Serialized.And []) = error "deserialize: Malformed `And` (should have at least two elements)"
```
Comment (GeorgSchneider) [inline daml/ContingentClaims/Claim.daml:126]: you say "should have at least two elements" but you handle single elements above gracefully :)

**Review by GeorgSchneider — APPROVED** [review]

File: daml/ContingentClaims/Claim.daml (hunk 5 above)
Comment (lucianojoublanc-da) [reply daml/ContingentClaims/Claim.daml:126]: yes that's true ... what this means is the identity `serialize . deserialize == identity` holds, but `deserialize . serialize /= identity` in the case where you have a `And [c]` single element node (it will reduce to a `c`)
